### PR TITLE
Add cache to AddressFactory::addressExists()

### LINKED ISSUE #26754: Too many queries to test if address_id exists
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

When there is a product in the cart and I'm not logged in, Prestashop is repeating this query 30 times

    SELECT SQL_NO_CACHE `id_address`  FROM ps_address a  WHERE a.`id_address` = 0 LIMIT 1

Additionally the logic is suspect, because the query does not check if the address is marked as deleted or if it is active. This would be necessary, since the addresses are soft-deleted in the address table.

The screenshot are from Profiling mode on a product page. Checking the individual queries in Stopwatch SQL shows that all queries are exactly the same.

### Expected behavior

Never check if address with `id_address = 0` exists. It must never exist.
Check only once if an address exists, remember the answer.
Even better: good execution logic would test only once for the existence of the address saved in the cart.

To-do:

- [x] #28321

- [ ]  Investigate the core of the issue to understand: why we are doing so much queries (related to how we store the adress in the cache of the cart) ?

### Steps to reproduce

1. Enable Profiling mode
3. Add a product to your cart
4. Scroll down to query doubles, you'll notice that there is 30x the same query on the table `address`.

### PrestaShop version(s) where the bug happened

1.7.8.0

### PHP version(s) where the bug happened

7.3

### If your bug is related to a module, specify its name and its version

_No response_

### PR DESCRIPTION
**Pull request by eternoendless**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 1.7.8.x
| Description?      | Implement cache option to avoid repeat requests
| Type?             | bug fix
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #26754
| Related PRs       | N/A
| How to test?      | Described here: https://github.com/PrestaShop/PrestaShop/issues/26754#issuecomment-1152941414
| Possible impacts? | Unknown

### CODE REVIEW

**Review by Progi1984 — APPROVED** [review]

File: classes/Cart.php [hunk 1]
```diff
@@ -2404,7 +2404,7 @@ public function getProductAddressId($product)
         }
 
         // Get delivery address of the product from the cart
-        if (!$this->addressFactory->addressExists($addressId)) {
+        if (!$this->addressFactory->addressExists($addressId, true)) {
```
Comment (PululuK) [inline classes/Cart.php:2407]: Make sure that the cache is cleaned when address is deleted.

FYI : https://github.com/PrestaShop/PrestaShop/pull/21616

Comment (eternoendless) [reply classes/Cart.php:2407]: This is done by Address automatically.

**Comment by Hlavtox** [comment]
Can we left join ps_address and ps_customer when creating Cart object? We will immediately see if all addresses and customer exists, reset it to zero if needed and eliminate all of these calls.

**Comment by eternoendless** [comment]
In PrestaShop's schema, one order is delivered to one address. From my point of view, the Cart - product - address relationship makes no sense and should be removed. But that's a whole different refactor league...

**Review by jolelievre — APPROVED** [review]
Thanks @eternoendless

**Comment by jolelievre** [comment]
@eternoendless it's weird the sanity tests keep failing even after restarting them, maybe your modification has unexpected impacts

**Comment by eternoendless** [comment]
It's because one of the modules being pushed on install crashes the Orders page when in debug mode. This should be fixed soon (the nightlies are failing for the same reason)

**Review by ghost — APPROVED** [review]

**Comment by Matt75** [comment]
> From my point of view, the Cart - product - address relationship makes no sense and should be removed. But that's a whole different refactor league...

It's related to the legacy multishipping feature, based on package system introduced in PrestaShop 1.5... remove this relationship will breaks a lot of modules that's restore multishipping feature.

In real world it can happens thats products into cart cannot been send with same carrier. For example some product are oversized etc...

It's a real need to be able to split a cart in multiple package, for exemple on Amazon, your cart can be separate in different package and sent using different carriers or if you're are a marketplace, products can be shipped from different warehouse...

Currently feature has been partially removed, because you can still see a cart separated to multiple orders currently but you cannot manage it correctly.
https://github.com/PrestaShop/prestashop-specs/issues/111
https://github.com/PrestaShop/docs/issues/508

**Comment by eternoendless** [comment]
Yep @Matt75 I remember your earlier comment about that. But I think the feature should be fully removed, and if necessary, be reintroduced differently. For example, via `ps_product_carrier`, or some other predefined relationship (to define carrier constraints), then recorded via separate orders. Definitely not within the cart.

**Comment by Matt75** [comment]
@eternoendless Yes, that's why I ask regularly for cases studies on my opened issues on this topic...

Current situation is not reliable for us as payment module developer because the core performs actions unwanted and we cannot control them.

Workflow is incomplete, so merchant installs modules to restore some removed features that's not cover all their use cases and on our side it's still a mess for payment module...

But we cannot change this without purpose a way for merchants to have this kind of features thanks to module cause top merchants use and needs this...

sorry for the digression 🤭

**Review by AureRita — APPROVED** [review]
Hi @eternoendless 

Thank you for your PR, I tested it and it seems to works, as we can see : 
image

I also launched an automated test on it, just in case, and it didn't find issues

**Comment by matthieu-rolland** [comment]
thank you @eternoendless and @AureRita 👍

**Pull request merged**